pages/api/applications/[applicationId]/stages/[stageId].ts

```diff
@@ -1,5 +1,215 @@
+import { isValidDate } from '../../../../../utils/date/validations';
+import { createJsonResponse, HttpMethod, HttpStatus, rejectHttpMethod } from '../../../../../utils/http/httpHelpers';
+import { withPrismaErrorHandling } from '../../../../../utils/prisma/prismaHelpers';
+import { ApiResponse, EmptyPayload, StatusMessageType } from '../../../../../types/apiResponse';
+import { Nullable } from '../../../../../types/utils';
+import { isInteger } from '../../../../../utils/numbers/validations';
 
-export default function handler(req: NextApiRequest, res: NextApiResponse) {
-  res.status(200).end(`Request sent to ${req.url}\n`);
+enum MessageType {
+  APPLICATION_STAGE_DELETE_UNAUTHORIZED,
+  APPLICATION_STAGE_DELETED_SUCCESSFULLY,
+  APPLICATION_STAGE_NOT_FOUND,
+  APPLICATION_STAGE_UPDATE_UNAUTHORIZED,
+  APPLICATION_STAGE_UPDATED_SUCCESSFULLY,
+  INVALID_APPLICATION_ID,
+  INVALID_APPLICATION_STAGE_ID,
+  INVALID_APPLICATION_STAGE_TYPE,
+  INVALID_DATE,
+  INVALID_EMOJI_UNICODE_HEX,
 }
+
+const prisma = new PrismaClient();
+
+const messages = Object.freeze({
+  [MessageType.APPLICATION_STAGE_DELETE_UNAUTHORIZED]: {
+    type: StatusMessageType.ERROR,
+    message: 'Application stage cannot be deleted by the user.',
+  },
+  [MessageType.APPLICATION_STAGE_DELETED_SUCCESSFULLY]: {
+    type: StatusMessageType.SUCCESS,
+    message: 'Application stage deleted successfully.',
+  },
+  [MessageType.APPLICATION_STAGE_NOT_FOUND]: {
+    type: StatusMessageType.ERROR,
+    message: 'Application stage cannot be found.',
+  },
+  [MessageType.APPLICATION_STAGE_UPDATE_UNAUTHORIZED]: {
+    type: StatusMessageType.ERROR,
+    message: 'Application stage cannot be updated by the user.',
+  },
+  [MessageType.APPLICATION_STAGE_UPDATED_SUCCESSFULLY]: {
+    type: StatusMessageType.SUCCESS,
+    message: 'Application stage updated successfully.',
+  },
+  [MessageType.INVALID_APPLICATION_ID]: {
+    type: StatusMessageType.ERROR,
+    message: 'Application id is invalid',
+  },
+  [MessageType.INVALID_APPLICATION_STAGE_ID]: {
+    type: StatusMessageType.ERROR,
+    message: 'Application stage id is invalid',
+  },
+  [MessageType.INVALID_APPLICATION_STAGE_TYPE]: {
+    type: StatusMessageType.ERROR,
+    message: 'Application stage type is invalid.',
+  },
+  [MessageType.INVALID_DATE]: { type: StatusMessageType.ERROR, message: 'Application stage date is invalid.' },
+  [MessageType.INVALID_EMOJI_UNICODE_HEX]: {
+    type: StatusMessageType.ERROR,
+    message: 'Application stage emoji unicode hex is invalid.',
+  },
+});
+
+function handler(userId: string, req: NextApiRequest, res: NextApiResponse) {
+  const method = req.method;
+  switch (method) {
+    case HttpMethod.PATCH:
+      handlePatch(userId, req, res);
+      break;
+    case HttpMethod.DELETE:
+      handleDelete(userId, req, res);
+      break;
+    default:
+      rejectHttpMethod(res, method);
+  }
+}
+
+async function handlePatch(
```

Comment:
Alright I found the issue (related to #28). So these are the fixes, in order:

1. The `handler()` function returns before the async is completed: merge #30 into this branch pls
2. The middleware from `jwtHelpers` and `prismaHelpers` return before the async is handled: merge #28 into main pls
3.  The Prisma helper didn't account for this as a bad parameter rather than an internal server error: merge #28 into main pls.

For (1), I think we need to let everyone know that if you are using async handlers, make sure your entire function finishes before you return.